kernel/read.c

```diff
@@ -53,37 +53,71 @@ static int EnsureExtensionHandle(PDEVICE_EXTENSION extension)
 }
 
 
+// Method I.
 ULONG PhysicalMemoryPartialRead(IN PDEVICE_EXTENSION extension,
                                       LARGE_INTEGER offset, unsigned char * buf,
                                       ULONG count) 
 {
-  ULONG page_offset = offset.QuadPart % PAGE_SIZE;
-  ULONG to_read = min(PAGE_SIZE - page_offset, count);
-  PUCHAR mapped_buffer = NULL;
-  SIZE_T ViewSize = PAGE_SIZE;
-  NTSTATUS NtStatus;
+	ULONG page_offset = offset.QuadPart % PAGE_SIZE;
+	ULONG to_read = min(PAGE_SIZE - page_offset, count);
+	PUCHAR mapped_buffer = NULL;
+	SIZE_T ViewSize = PAGE_SIZE;
+	NTSTATUS NtStatus;
+	ULONG result = 0;
 
-  if (EnsureExtensionHandle(extension)) 
-  {
-    /* Map page into the Kernel AS */
-    NtStatus = ZwMapViewOfSection(extension->MemoryHandle, (HANDLE) -1,
+	if (!(openPhysMemSectionHandle(extension)))
+	{
+		DbgPrint("Error: physical device handle not available!\n"); // real error
+		return 0;
+	}
+		
+	// The mapview should never fail...
+	NtStatus = ZwMapViewOfSection(extension->MemoryHandle, (HANDLE) -1,
 				  &mapped_buffer, 0L, PAGE_SIZE, &offset,
 				  &ViewSize, ViewUnmap, 0, PAGE_READONLY);
+				  
+	if ((NtStatus != STATUS_SUCCESS) || (!mapped_buffer))
+	{
+		DbgPrint("Error: ZwMapViewOfSection failed. Offset 0x%llX, status %08x.\n", offset.QuadPart,NtStatus); // real error
+		return 0;
+	}
+	
+	// ... but reading from it may.
+	
+	
+	// By the way, is that right that we create a mapview section for each tiny read?

```

Comment:
Yes, but only since my updates for large read support. And **only** if there is no VSM/Hyper-V layer around. I struggle finding barebone PCs that can do the large reads. It's everywhere. 

Meaning: we might assume small reads being more frequent.